cli/start.go

```diff
@@ -244,11 +244,99 @@ func MakeStartCommand() *cobra.Command {
 	cmd.PersistentFlags().Bool(
 		"development",
 		cfg.GetBool(configFlags["development"]),
-		"Enables a set of features that make development easier but should not be enabled in production",
+		developmentDescription,
 	)
 	cmd.Flags().Bool(
 		"no-encryption",
 		cfg.GetBool(configFlags["no-encryption"]),
 		"Skip generating an encryption key. Encryption at rest will be disabled. WARNING: This cannot be undone.")
 	return cmd
 }
+
+func getOrCreateEncryptionKey(kr keyring.Keyring, cfg *viper.Viper, opts []node.Option) ([]node.Option, error) {
+	encryptionKey, err := kr.Get(encryptionKeyName)
+	if err != nil && errors.Is(err, keyring.ErrNotFound) && !cfg.GetBool("datastore.noencryption") {
+		encryptionKey, err = crypto.GenerateAES256()
+		if err != nil {
+			return nil, err
+		}
+		err = kr.Set(encryptionKeyName, encryptionKey)
+		if err != nil {
+			return nil, err
+		}
+		log.Info("generated encryption key")
+	} else if err != nil && !errors.Is(err, keyring.ErrNotFound) {
+		return nil, err
+	}
+	opts = append(opts, node.WithBadgerEncryptionKey(encryptionKey))
+	return opts, nil
+}
+
+func getOrCreatePeerKey(kr keyring.Keyring, opts []node.Option) ([]node.Option, error) {
+	peerKey, err := kr.Get(peerKeyName)
+	if err != nil && errors.Is(err, keyring.ErrNotFound) {
+		peerKey, err = crypto.GenerateEd25519()
+		if err != nil {
+			return nil, err
+		}
+		err = kr.Set(peerKeyName, peerKey)
+		if err != nil {
+			return nil, err
+		}
+		log.Info("generated peer key")
+	} else if err != nil {
+		return nil, err
+	}
+	return append(opts, net.WithPrivateKey(peerKey)), nil
+}
+
+func getOrCreateIdentity(kr keyring.Keyring, opts []node.Option) ([]node.Option, error) {
+	identityBytes, err := kr.Get(nodeIdentityKeyName)
+	if err != nil {
+		if !errors.Is(err, keyring.ErrNotFound) {
+			return nil, err
+		}
+		privateKey, err := crypto.GenerateSecp256k1()
+		if err != nil {
+			return nil, err
+		}
+		identityBytes := privateKey.Serialize()
+		err = kr.Set(nodeIdentityKeyName, identityBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	nodeIdentity, err := identity.FromPrivateKey(
+		secp256k1.PrivKeyFromBytes(identityBytes),
+		time.Duration(0),
+		immutable.None[string](),
+		immutable.None[string](),
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(opts, db.WithNodeIdentity(nodeIdentity)), nil
+}
+
+func addEphemeralIdentity(opts []node.Option) ([]node.Option, error) {
```

Comment:
todo: 'ephemeral' suggests that this is temporary, however I cannot spot any documentation as to why it is so, and it seems a little odd if it is only kept in memory for a dev mode database (potentially file-based) - surely things will break if it is restarted?

I'm happy with a ticket being opened and linked with a code-comment-todo if the restart problem is real, as this feels quite niche.  But please make sure that this function is documented.

Might be a bit of a hassle now, but long term we could perhaps just store this in plain text on disk (it is in dev-mode).

Reply:
well, it IS temporary just to have some identity to play around with during development. I didn't think it was important to persist it.

Created a task https://github.com/sourcenetwork/defradb/issues/3148

Reply:
Dev/test environments can persist for quite a while, it is not uncommon to have permanent environments for that.  At one org I worked for each dev had their own permanent, always running, dev deployment (web server, plus infrastructure).